### PR TITLE
feat(codeguruprofiler): ProfilingGroup

### PR DESCRIPTION
**Pull request by SeekerWing**
fixes https://github.com/aws/aws-cdk/issues/6984 by creating L2 construct and functions to allow for policies to be assigned to execution roles.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-codeguruprofiler/lib/profiling-group.ts [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+/**
+ * Properties for creating a new Profiling Group.
+ */
+export interface ProfilingGroupProps {
+
+  /**
+   * A name for the profiling group.
+   */
+  readonly profilingGroupName: string;
```
Comment (eladb) [inline packages/@aws-cdk/aws-codeguruprofiler/lib/profiling-group.ts:13]: Can this be omitted? Can we produce a sensible default here?

File: packages/@aws-cdk/aws-codeguruprofiler/lib/profiling-group.ts [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+}
+
+/**
+ * A new Profiling Group.
+ */
+export class ProfilingGroup extends Resource {
```
Comment (eladb) [inline packages/@aws-cdk/aws-codeguruprofiler/lib/profiling-group.ts:20]: Please implement `IProfilingGroup` which extends `IResource`

File: packages/@aws-cdk/aws-codeguruprofiler/lib/profiling-group.ts [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+  /**
+   * The name of this ProfilingGroup
+   */
+  public readonly profilingGroupName: string;
+
+  /**
+   * The ARN of this ProfilingGroup
+   */
+  public readonly profilingGroupArn: string;
+
```
Comment (eladb) [inline packages/@aws-cdk/aws-codeguruprofiler/lib/profiling-group.ts:31]: Missing `ProfilingGroup.fromProfilingGroupName` and `fromProfilingGroupArn`

File: packages/@aws-cdk/aws-codeguruprofiler/README.md [hunk 4]
```diff
@@ -9,8 +9,28 @@
 ---
 <!--END STABILITY BANNER-->
 
-This module is part of the [AWS Cloud Development Kit](https://github.com/aws/aws-cdk) project.
+Amazon CodeGuru Profiler collects runtime performance data from your live applications, and provides recommendations that can help you fine-tune your application performance. Using machine learning algorithms, CodeGuru Profiler can help you find your most expensive lines of code and suggest ways you can improve efficiency and remove CPU bottlenecks. Amazon CodeGuru Profiler provides different visualizations of profiling data to help you identify what code is running on the CPU, see how much time is consumed, and suggest ways to reduce CPU utilization.
```
Comment (eladb) [inline packages/@aws-cdk/aws-codeguruprofiler/README.md:12]: limit to 100 chars

File: packages/@aws-cdk/aws-codeguruprofiler/lib/profiling-group.ts (hunk 1 above)
Comment (SeekerWing) [reply packages/@aws-cdk/aws-codeguruprofiler/lib/profiling-group.ts:13]: i've changed the approach to the following, not an expert by any means, let me know your thoughts.

```ts
physicalName: props.profilingGroupName || Lazy.stringValue({ produce: () => this.node.uniqueId })
```

File: packages/@aws-cdk/aws-codeguruprofiler/lib/profiling-group.ts (hunk 3 above)
Comment (SeekerWing) [reply packages/@aws-cdk/aws-codeguruprofiler/lib/profiling-group.ts:31]: not very experienced with the implementation of such methods, i've given it my best guess.

File: packages/@aws-cdk/aws-codeguruprofiler/README.md (hunk 4 above)
Comment (SeekerWing) [reply packages/@aws-cdk/aws-codeguruprofiler/README.md:12]: reduced to 2 sentences, hope that's okay.

File: packages/@aws-cdk/aws-codeguruprofiler/lib/profiling-group.ts [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    class Import extends ProfilingGroupBase {
+      public readonly profilingGroupName = profilingGroupName;
+      public readonly profilingGroupArn = stack.formatArn({
+        service: 'codeguru-profiler',
+        resource: 'profilingGroup',
+        resourceName: profilingGroupName,
+      });
+    }
+
+    return new Import(scope, id);
```
Comment (eladb) [inline packages/@aws-cdk/aws-codeguruprofiler/lib/profiling-group.ts:35]: Delegate to `fromProfilingGroupArn`

File: packages/@aws-cdk/aws-codeguruprofiler/package.json [hunk 6]
```diff
@@ -67,14 +67,18 @@
   "devDependencies": {
     "@aws-cdk/assert": "0.0.0",
     "cdk-build-tools": "0.0.0",
+    "cdk-integ-tools": "0.0.0",
     "cfn2ts": "0.0.0",
+    "nodeunit": "^0.11.3",
```
Comment (eladb) [inline packages/@aws-cdk/aws-codeguruprofiler/package.json:72]: Please use jest (nodeunit is deprecated)

File: packages/@aws-cdk/aws-codeguruprofiler/test/integ.profiler-group.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
+import { AccountRootPrincipal, Role } from '@aws-cdk/aws-iam';
+import { App, Stack, StackProps } from '@aws-cdk/core';
+import { ProfilingGroup } from '../lib/profiling-group';
```
Comment (eladb) [inline packages/@aws-cdk/aws-codeguruprofiler/test/integ.profiler-group.ts:3]: This way you can verify that you indeed export the types:

```suggestion
import { ProfilingGroup } from '../lib';
```

File: packages/@aws-cdk/aws-codeguruprofiler/test/profiling-group.test.ts [hunk 8]
```diff
@@ -0,0 +1,376 @@
+import { expect } from '@aws-cdk/assert';
+import { AccountRootPrincipal, Role } from '@aws-cdk/aws-iam';
+import { Stack } from '@aws-cdk/core';
+import { ProfilingGroup } from '../lib/profiling-group';
```
Comment (eladb) [inline packages/@aws-cdk/aws-codeguruprofiler/test/profiling-group.test.ts:4]:
```suggestion
import { ProfilingGroup } from '../lib';
```

File: packages/@aws-cdk/aws-codeguruprofiler/lib/profiling-group-base.ts [hunk 9]
```diff
@@ -0,0 +1,81 @@
+import { Grant, IGrantable } from '@aws-cdk/aws-iam';
```
Comment (eladb) [inline packages/@aws-cdk/aws-codeguruprofiler/lib/profiling-group-base.ts:1]: No need to split into a separate file.

File: packages/@aws-cdk/aws-codeguruprofiler/lib/profiling-group-base.ts [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+   *  - codeguru-profiler:GetProfile
+   *  - codeguru-profiler:DescribeProfilingGroup
+   *
+   * @param grantee Principal to grant read rights to
+   */
+  grantRead(grantee: IGrantable): void;
+
+}
+
+export abstract class ProfilingGroupBase extends Resource implements IProfilingGroup {
```
Comment (eladb) [inline packages/@aws-cdk/aws-codeguruprofiler/lib/profiling-group-base.ts:37]: The base type should not be exported (jsii knows how to erase it)

File: packages/@aws-cdk/aws-codeguruprofiler/README.md [hunk 11]
```diff
@@ -9,8 +9,28 @@
...
+
+```ts
+// The execution role of your application that publishes to the ProfilingGroup via CodeGuru Profiler Profiling Agent. (the following is merely an example)
+const publishAppRole = new Role(stack, 'PublishAppRole', {
+  assumedBy: new AccountRootPrincipal(),
+});
+
+const profilingGroup = new ProfilingGroup(stack, 'MyProfilingGroup', {
+  profilingGroupName: 'MyAwesomeProfilingGroup',
+});
```
Comment (eladb) [inline packages/@aws-cdk/aws-codeguruprofiler/README.md:34]: Prefer the simplest example:

```suggestion
const profilingGroup = new ProfilingGroup(stack, 'MyProfilingGroup');
```

File: packages/@aws-cdk/aws-codeguruprofiler/lib/profiling-group.ts [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+  }
+
+  public readonly profilingGroupName: string;
+  public readonly profilingGroupArn: string;
+
+  constructor(scope: Construct, id: string, props: ProfilingGroupProps) {
+    super(scope, id, {
+      physicalName: props.profilingGroupName || Lazy.stringValue({ produce: () => this.node.uniqueId }),
```
Comment (eladb) [inline packages/@aws-cdk/aws-codeguruprofiler/lib/profiling-group.ts:47]:
```suggestion
      physicalName: props.profilingGroupName ?? Lazy.stringValue({ produce: () => this.node.uniqueId }),
```

File: packages/@aws-cdk/aws-codeguruprofiler/test/integ.profiler-group.ts [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+
+class ProfilerGroupIntegrationTest extends Stack {
+  constructor(scope: App, id: string, props?: StackProps) {
+    super(scope, id, props);
+
+    const profilingGroup = new ProfilingGroup(this, 'MyProfilingGroup', {
+      profilingGroupName: 'MyAwesomeProfilingGroup',
+    });
```
Comment (eladb) [inline packages/@aws-cdk/aws-codeguruprofiler/test/integ.profiler-group.ts:11]: Check that the default works:

```suggestion
    const profilingGroup = new ProfilingGroup(this, 'MyProfilingGroup');
```

File: packages/@aws-cdk/aws-codeguruprofiler/lib/profiling-group.ts (hunk 12 above)
Comment (eladb) [reply packages/@aws-cdk/aws-codeguruprofiler/lib/profiling-group.ts:47]: Also - are there any constraints on the profile group name that we should take into account when using unqiueId?

**Pull request merged**